### PR TITLE
Feat add _etag for object storage

### LINKED ISSUE #65312: Add columns `_etag`, `_last_modified`, `_storage_class` for `S3` storage.
See https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html
Similar to `_size`.

### PR DESCRIPTION
**Pull request by skyoct**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `_etag` virtual column for S3 table engine. Fixes #65312.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 766130bc98c116d198343f8fee6e0e5527fad712 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65386/766130bc98c116d198343f8fee6e0e5527fad712/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by skyoct** [comment]
@kssenii Hi, How can I rerun CI?

**Comment by kssenii** [comment]
CI fails for a reason.

> ClickHouse build check — 34/35 artifact groups are OK 

If you open `details`, you'll see:
```
clickhouse_common_io.dir/IO/WriteBufferFromS3.cpp.o -c /build/src/IO/WriteBufferFromS3.cpp
Jun 21 02:43:34 /build/src/IO/S3/getObjectInfo.h:18:12: error: redundant string initialization [readability-redundant-string-init,-warnings-as-errors]
Jun 21 02:43:34    18 |     String etag = "";
Jun 21 02:43:34       |            ^~~~~~~~~
Jun 21 02:43:34       |            etag
Jun 21 02:43:34 48866 warnings generated.
```

**Review by kssenii — COMMENTED** [review]
Could you please add `_etag` to a test? See https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/02245_s3_virtual_columns.sql

**Comment by skyoct** [comment]
@kssenii Ok, Thank you very much.

**Comment by skyoct** [comment]
@kssenii The etag should just be a random string, and I have no way of knowing its value. Can I just check if it is empty here?

**Comment by kssenii** [comment]
> @kssenii The etag should just be a random string, and I have no way of knowing its value. Can I just check if it is empty here?

Yes, of course.

**Comment by skyoct** [comment]
image
@kssenii  Hi, It seems that I can't find what caused the error from this.

**Comment by kssenii** [comment]
> Stateless tests (release) — fail: 1

00992_system_parts_race_condition_zookeeper_long


-- this test is flaky, do not mind this failure. I will mark this failure is ignored. The tests will continue once I fix `Cloud fork sync (only for ClickHouse Inc. employees) — tests failed` -- will do now.

**Comment by kssenii** [comment]
> Docs check — Found errors in docs 

Please check this failure it should be related to changes

**Comment by kssenii** [comment]
@skyoct, could you please fix the build?

**Comment by skyoct** [comment]
@kssenii Okay, I will fix it immediately.

**Comment by kssenii** [comment]
> Cloud fork sync (only for ClickHouse Inc. employees) — Failed to merge into the sync PR

Fixed it now, @skyoct please don't do merge with master for this PR now, if it is not necessary, because it now breaks the sync each time for this PR.

Could you please also fix style-check?
```
In /ClickHouse/tests/queries/0_stateless/02270_errors_in_files.sh line 31:
rm "${user_files_path}"/test_02270_1.csv
    ^----------------^ SC2154 (warning): user_files_path is referenced but not assigned (did you mean 'USER_FILES_PATH'?).
```

File: tests/queries/0_stateless/02270_errors_in_files.sh [hunk 1]
```diff
@@ -25,9 +25,10 @@ ${CLICKHOUSE_CLIENT} --query "INSERT INTO TABLE FUNCTION file('test_02270_2.csv.
...
 ${CLICKHOUSE_CLIENT} --query "SELECT * FROM file('test_02270*.csv.gz', 'CSV', 'a String, b String')" --input_format_parallel_parsing 0 2>&1 | grep -o -m1 "test_02270_2.csv.gz"
 
-rm -f "${CLICKHOUSE_TMP}"/test_02270_1.csv
-rm -f "${CLICKHOUSE_TMP}"/test_02270_2.csv
-rm -f "${USER_FILES_PATH}"/test_02270_1.csv
-rm -f "${USER_FILES_PATH}"/test_02270_2.csv
-rm -f "${USER_FILES_PATH}"/test_02270_1.csv.gz
-rm -f "${USER_FILES_PATH}"/test_02270_2.csv.gz
+# remove files
+rm "${CLICKHOUSE_TMP}"/test_02270_1.csv
```
Comment (kssenii) [inline tests/queries/0_stateless/02270_errors_in_files.sh:29]: btw why change this?

Comment (skyoct) [reply tests/queries/0_stateless/02270_errors_in_files.sh:29]: @kssenii I feel that this test case not passing is quite strange.  Want to try again.

**Comment by skyoct** [comment]
image
@kssenii Is it necessary to rerun the tests after this change?

**Comment by skyoct** [comment]
> image @kssenii Is it necessary to rerun the tests after this change?

@kssenii Can you help me with this?

**Comment by kssenii** [comment]
> @kssenii Hi, Can you help review?

Sorry, Cloud For Sync gets stuck in pending state for this PR, I restarted it yesterday via github api, but it does not finish still. I will try again to reset it again via github api, and if it does not help again - I will push an empty commit.

**Comment by alexey-milovidov** [comment]
@kssenii, it is very strange to have ETag in double quotes:

```
:) SELECT _path, _etag FROM s3('s3://ookla-open-data/parquet/performance/type=*/year=*/quarter=*/*.parquet') LIMIT 10

SELECT
    _path,
    _etag
FROM s3('s3://ookla-open-data/parquet/performance/type=*/year=*/quarter=*/*.parquet')
LIMIT 10

Query id: 992f4385-b9e1-4194-8f3a-55843b093615

    ┌─_path─────────────────────────────────────────────────────────────────────────────────────────────────────────┬─_etag─────────────────────────────────┐
 1. │ ookla-open-data/parquet/performance/type=fixed/year=2019/quarter=1/2019-01-01_performance_fixed_tiles.parquet │ "921e7e414a81f5eb426810e73a3ee108-28" │
 2. │ ookla-open-data/parquet/performance/type=fixed/year=2019/quarter=1/2019-01-01_performance_fixed_tiles.parquet │ "921e7e414a81f5eb426810e73a3ee108-28" │
 3. │ ookla-open-data/parquet/performance/type=fixed/year=2019/quarter=1/2019-01-01_performance_fixed_tiles.parquet │ "921e7e414a81f5eb426810e73a3ee108-28" │
 4. │ ookla-open-data/parquet/performance/type=fixed/year=2019/quarter=1/2019-01-01_performance_fixed_tiles.parquet │ "921e7e414a81f5eb426810e73a3ee108-28" │
 5. │ ookla-open-data/parquet/performance/type=fixed/year=2019/quarter=1/2019-01-01_performance_fixed_tiles.parquet │ "921e7e414a81f5eb426810e73a3ee108-28" │
 6. │ ookla-open-data/parquet/performance/type=fixed/year=2019/quarter=1/2019-01-01_performance_fixed_tiles.parquet │ "921e7e414a81f5eb426810e73a3ee108-28" │
 7. │ ookla-open-data/parquet/performance/type=fixed/year=2019/quarter=1/2019-01-01_performance_fixed_tiles.parquet │ "921e7e414a81f5eb426810e73a3ee108-28" │
 8. │ ookla-open-data/parquet/performance/type=fixed/year=2019/quarter=1/2019-01-01_performance_fixed_tiles.parquet │ "921e7e414a81f5eb426810e73a3ee108-28" │
 9. │ ookla-open-data/parquet/performance/type=fixed/year=2019/quarter=1/2019-01-01_performance_fixed_tiles.parquet │ "921e7e414a81f5eb426810e73a3ee108-28" │
10. │ ookla-open-data/parquet/performance/type=fixed/year=2019/quarter=1/2019-01-01_performance_fixed_tiles.parquet │ "921e7e414a81f5eb426810e73a3ee108-28" │
    └───────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────────────────────────┘
```

It looks like a bug in AWS SDK.

**Comment by alexey-milovidov** [comment]
And I don't see implementation for the URL engine, where it would also be natural.

**Comment by alexey-milovidov** [comment]
@kssenii, `_etag` does not filter files, and this is contrary to my expectations:

https://pastila.nl/?05873082/36052e9869445d6b588d02dc837b8be7#c8bUlv6de6qe1JNHqp389w==

**Pull request merged**